File: 04-07-2025_Find_the_K-th_Character_in_String_Game.cpp
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        long long a =1;
        int n=0;
        while(k>a){
            a = a*2;
            n++;
        }
        int m =0;
        while(n>0 && k>1){
            if(a/2 <k){
                a/=2;
                k-=a;
                if(operations[n-1]==1) m++;
            }
            else{
                a/=2;
            }
            n--;
        }
        return 'a'+(m%26);
    }
};

File: 04-07-2025_Find_the_K-th_Character_in_String_Game.py
class Solution:
    def kthCharacter(self, k: int, operations: list[int]) -> str:
        a = 1  # current length of word
        n = 0  # number of operations needed to reach length ≥ k

        while k > a:
            a *= 2
            n += 1

        m = 0  # count of type-1 (shift) operations that affect the k-th character

        while n > 0 and k > 1:
            if a // 2 < k:
                a //= 2
                k -= a
                if operations[n - 1] == 1:
                    m += 1
            else:
                a //= 2
            n -= 1

        return chr(ord('a') + (m % 26))
